src/bulk_insert/bulk_insert.c

```diff
@@ -150,70 +150,104 @@ static int _BulkInsert_ProcessFile(GraphContext *gc, const char *data,
-				GraphEntity_AddProperty(ge, prop_indices[i], value);
-			}
+	while(data_idx < data_len) {
+		Node n;
+		GraphEntity *ge;
+		Graph_CreateNode(gc->g, label_id, &n);
+		ge = (GraphEntity *)&n;
+		// process entity attributes
+		for(uint i = 0; i < prop_count; i++) {
+			SIValue value = _BulkInsert_ReadProperty(data, &data_idx);
+			// skip invalid attribute values
+			if(!(SI_TYPE(value) & SI_VALID_PROPERTY_VALUE)) continue;
+			GraphEntity_AddProperty(ge, prop_indices[i], value);
 		}
 	}
 
+	Graph_SetMatrixPolicy(gc->g, SYNC_POLICY_RESIZE);
+	if(prop_indices) rm_free(prop_indices);
+	return BULK_OK;
+}
+
+static int _BulkInsert_ProcessEdgeFile(GraphContext *gc, const char *data,
+								   size_t data_len, SchemaType type) {
+
+	int label_id;
+	uint prop_count;
+	size_t data_idx = 0;
+
+	// read the CSV file header
+	// and commit all labels and properties it introduces
+	Attribute_ID *prop_indices = _BulkInsert_ReadHeader(gc, type, data,
+			&data_idx, &label_id, &prop_count);
+
+	// sync matrix once
+	Graph_GetRelationMatrix(gc->g, label_id, false);
+	Graph_GetAdjacencyMatrix(gc->g, false);
+	Graph_SetMatrixPolicy(gc->g, SYNC_POLICY_NOP);
+
 	//--------------------------------------------------------------------------
 	// load edges
 	//--------------------------------------------------------------------------
 
-	if(type == SCHEMA_EDGE) {
-		while(data_idx < data_len) {
-			Edge e;
-			GraphEntity *ge;
-
-			// next 8 bytes are source ID
-			NodeID src = *(NodeID *)&data[data_idx];
-			data_idx += sizeof(NodeID);
-			// next 8 bytes are destination ID
-			NodeID dest = *(NodeID *)&data[data_idx];
-			data_idx += sizeof(NodeID);
-
-			Graph_CreateEdge(gc->g, src, dest, label_id, &e);
-			ge = (GraphEntity *)&e;
-
-			// process entity attributes
-			for(uint i = 0; i < prop_count; i++) {
-				SIValue value = _BulkInsert_ReadProperty(data, &data_idx);
-				// skip invalid attribute values
-				if(!(SI_TYPE(value) & SI_VALID_PROPERTY_VALUE)) continue;
-				GraphEntity_AddProperty(ge, prop_indices[i], value);
-			}
+	while(data_idx < data_len) {
+		Edge e;
+		GraphEntity *ge;
+
+		// next 8 bytes are source ID
+		NodeID src = *(NodeID *)&data[data_idx];
+		data_idx += sizeof(NodeID);
+		// next 8 bytes are destination ID
+		NodeID dest = *(NodeID *)&data[data_idx];
+		data_idx += sizeof(NodeID);
+
+		Graph_CreateEdge(gc->g, src, dest, label_id, &e);
+		ge = (GraphEntity *)&e;
+
+		// process entity attributes
+		for(uint i = 0; i < prop_count; i++) {
+			SIValue value = _BulkInsert_ReadProperty(data, &data_idx);
+			// skip invalid attribute values
+			if(!(SI_TYPE(value) & SI_VALID_PROPERTY_VALUE)) continue;
+			GraphEntity_AddProperty(ge, prop_indices[i], value);
 		}
 	}
 
+	Graph_SetMatrixPolicy(gc->g, SYNC_POLICY_RESIZE);
 	if(prop_indices) rm_free(prop_indices);
 	return BULK_OK;
 }
 
-static int _BulkInsert_ProcessTokens(GraphContext *gc, int token_count,
+static int _BulkInsert_ProcessNodeTokens(GraphContext *gc, int token_count,
```

Comment:
`_BulkInsert_ProcessEdgeTokens` and `_BulkInsert_ProcessNodeTokens` are practically the same consolidate
and use `type` to determine which function to call `_BulkInsert_ProcessNodeFile` or `_BulkInsert_ProcessEdgeFile`